libs/parser/parser.c

```diff
@@ -0,0 +1,425 @@
+
+
+parser_t *parser_alloc(int maxHeadersNb, int maxFieldsNb)
+{
+	parser_t *result;
+
+	if (parser_common.regexCompErr != 0) {
+		fprintf(stderr, "Parser: an error occurred during Regex compilation.\n");
+		return NULL;
+	}
+
+	if (maxHeadersNb <= 0 || maxFieldsNb <= 0) {
+		fprintf(stderr, "Parser: invalid arguments. Both of them have to be bigger than zero.\n");
+		return NULL;
+	}
+
+	if (SIZE_MAX / maxHeadersNb < sizeof(parser_headerInfo_t)) {
+		fprintf(stderr, "Parser: maximum header number exceeded limit value\n");
+		return NULL;
+	}
+
+	if (SIZE_MAX / maxFieldsNb < sizeof(parser_field_t)) {
+		fprintf(stderr, "Parser: maximum fields number exceeded limit value\n");
+		return NULL;
+	}
+
+	result = malloc(sizeof(parser_t));
+	if (result == NULL) {
+		fprintf(stderr, "Parser: malloc error\n");
+		return NULL;
+	}
+
+	result->headerInfos = malloc(sizeof(parser_headerInfo_t) * maxHeadersNb);
+	if (result->headerInfos == NULL) {
+		fprintf(stderr, "Parser: malloc error\n");
+		free(result);
+		return NULL;
+	}
+
+	result->headersMap = hmap_init(maxHeadersNb);
+	if (result->headersMap == NULL) {
+		fprintf(stderr, "Parser: error during hash map initialisation\n");
+		free(result->headerInfos);
+		free(result);
+		return NULL;
+	}
+
+	result->fields = malloc(sizeof(parser_field_t) * maxFieldsNb);
+	if (result->fields == NULL) {
+		fprintf(stderr, "Parser: malloc error\n");
+		hmap_free(result->headersMap);
+		free(result->headerInfos);
+		free(result);
+		return NULL;
+	}
+
+	result->fieldsMap = hmap_init(maxFieldsNb);
+	if (result->fieldsMap == NULL) {
+		fprintf(stderr, "Parser: error during hash map initialisation\n");
+		free(result->fields);
+		hmap_free(result->headersMap);
+		free(result->headerInfos);
+		free(result);
+		return NULL;
+	}
+
+	return result;
+}
+
+
+void parser_free(parser_t *p)
+{
+	if (p == NULL) {
+		return;
+	}
+
+	hmap_free(p->headersMap);
+	free(p->headerInfos);
+
+	hmap_free(p->fieldsMap);
+	free(p->fields);
+
+	free(p);
+}
+
+
+int parser_headerAdd(parser_t *p, const char *headerName, int (*converter)(const hmap_t *))
+{
+	unsigned int id;
+
+	if (p == NULL || headerName == NULL || converter == NULL) {
+		fprintf(stderr, "Parser: invalid arguments. Only non-NULL arguments are correct\n");
+		return -1;
+	}
+
+	/* Checking if there is enough memory for a new header */
+	if (p->headersMap->capacity == p->headersMap->size) {
+		fprintf(stderr, "Parser: maximum number of header was exceeded.\n");
+		return -1;
+	}
+
+	if (strlen(headerName) > MAX_HEADER_LEN) {
+		fprintf(stderr, "Parser: header name \"%s\" is too long\n", headerName);
+		return -1;
+	}
+
+	id = p->headersMap->size;
+
+	p->headerInfos[id].converter = converter;
+	strcpy(p->headerInfos[id].headerName, headerName);
+
+	if (hmap_insert(p->headersMap,
+			p->headerInfos[id].headerName,
+			&(p->headerInfos[id])) == -1) {
+		fprintf(stderr, "Parser: it is impossible to define two headers with the same names\n");
```

Comment:
Please use imperative/nominal sentences as error messages (if possible). They are always shorter and straight informative.
```suggestion
		fprintf(stderr, "Parser: header %s already exists!\n", p->headerInfos[id].headerName);
```